src/Compilers/Core/Portable/Text/SourceText.cs

```diff
@@ -633,7 +633,19 @@ public virtual SourceText WithChanges(IEnumerable<TextChange> changes)
                 // there can be no overlapping changes
                 if (change.Span.Start < position)
                 {
-                    throw new ArgumentException(CodeAnalysisResources.ChangesMustBeOrderedAndNotOverlapping, nameof(changes));
+                    // Handle the case of unordered changes by sorting the input and retrying. This is inefficient, but
+                    // downstream consumers have been known to hit this case in the past and we want to avoid crashes.
+                    // https://github.com/dotnet/roslyn/pull/26339
+                    if (change.Span.End <= changeRanges.Last().Span.Start)
```

Comment:
it's not clear ot me why this check is correct.  why is it sufficient to only check the last change range?

Reply:
I think the answer is because this is in a loop that goes through all of the changes.

---
In reply to: [187675569](https://github.com/dotnet/roslyn/pull/26798#discussion_r187675569) [](ancestors = 187675569)

Reply:
➡️ `changeRanges` is always sorted (a condition that was true both before and after this change).